[OCPERT-138]  Automate blocking sec-alerts detection in oar tool through new command `check-blocking-sec-alerts`

jhuttana opened the pull request:
Feature: Add blocking security alerts detection
    
    - Add worksheet section A29/B29 with dropdown (Yes/No) and conditional formatting
    - Implement automatic RHSA security alerts checking via errata API
    - Add 'check-blocking-sec-alerts' CLI command with --check-only, --output, --notify options
    - Auto-update worksheet and notify @{version}-qe-leads when blocking alerts found
    - Include comprehensive tests and documentation updates
    
    Enables automatic detection and team notification of blocking security alerts
    for RHSA advisories with real-time worksheet synchronization.

jhuttana commented:
> JFYI, this function(blocking secalert checking) is already supported in cmd `change-advisory-status`, because it is blocker for this task. notification is a little complicated, because different blocking secalerts have different contacts, you can refer to workseet [here](https://docs.google.com/spreadsheets/d/17NQZflrWaxgIEAHcmevC3f2jqKy7HYWarVq_CxGVHAc/edit?gid=0#gid=0), if we just wanna notify release lead, it's not necessary, it is just like a FYI message. if you wanna add a dedicated cmd to check blocking alert at early stage, you just need to use existing funcs in new cli cmd. ref: https://github.com/openshift/release-tests/blob/oar-errata-release-flow/oar/core/advisory.py#L186 ref: https://github.com/openshift/release-tests/blob/oar-errata-release-flow/oar/core/advisory.py#L671
> 
+1 for sending notification and using the  existing function `has_blocking_secruity_alert`.
> BTW, if you want to update blocking secalert status (y/n) in worksheet, you just need to update cell under col `others` in table `Critical Issue`
> 
I will give it a try and see how it looks when updated to `others` column
> /cc @tomasdavidorg

tomasdavidorg commented:
+1 to move the results of the sec. alerts check from B29. However what will be the result? Can we create a general jira issue because just yes/no option does not look good. We could use the issue as a reference if we send and email to prod team or notify the ART team. 

On the other side creating a new command would have its own item in the check list (but this would break the current template). 

I agree that checking the sec. alerts in the `change-advisory-status` is too late and we do it manually sooner and it would have deserved an earlier automatic check.

jhuttana commented:
I have addressed the review comments by @rioliu-rh and @ming1013 in the latest commit. 

The sample outputs from this commit
```
$ oar -r 4.18.20 check-blocking-sec-alerts -h
Usage: oar check-blocking-sec-alerts [OPTIONS]

  Check for blocking security alerts across all advisories and update
  worksheet (A29/B29) and Others column

Options:
  --check-only          Only check for blocking alerts without updating
                        worksheet
  --output [text|json]  Output format for results
  -h, --help            Show this message and exit.
```
```
$ oar -r 4.18.20 check-blocking-sec-alerts
2025-07-23T15:06:25Z: INFO: Checking for blocking security alerts across all advisories...
2025-07-23T15:08:14Z: INFO: RHSA advisory 152071 does not have blocking secalert
2025-07-23T15:08:21Z: INFO: RHSA advisory 152072 does not have blocking secalert
2025-07-23T15:08:21Z: INFO: 🔍 Security alert check complete: 4 total, 2 RHSA checked, 0 blocking, 0 errors
🔍 Blocking Security Alerts Check for 4.18.20
==================================================
📊 Summary:
  • Total advisories: 4
  • RHSA advisories checked: 2
  • Blocking advisories found: 0
  • Errors encountered: 0

✅ No blocking security alerts found

📋 Advisory Details:
  ✅ 152071 (RHSA/image)
  ✅ 152072 (RHSA/rpm)
  ⏭️ 152073 (RHBA/extras) (RHBA advisories don't have security alerts)
  ⏭️ 152074 (RHBA/metadata) (RHBA advisories don't have security alerts)
2025-07-23T15:08:21Z: INFO: Updating worksheet with blocking security alerts status...
📝 Worksheet Status:
  • Value remains 'No' in cell B29 (no change needed)

✅ Exit code 0: No blocking security alerts
2025-07-23T15:08:23Z: ERROR: check blocking security alerts failed
Traceback (most recent call last):
  File "/home/jhuttana/workarea/openShift_qe/release-tests/oar/cli/cmd_check_blocking_sec_alerts.py", line 193, in check_blocking_sec_alerts
    ctx.exit(0)
    ~~~~~~~~^^^
  File "/home/jhuttana/.local/lib/python3.13/site-packages/click/core.py", line 699, in exit
    raise Exit(code)
click.exceptions.Exit: 0
❌ Error: 0
```

jhuttana commented:
> +1 to move the results of the sec. alerts check from B29. However what will be the result? Can we create a general jira issue because just yes/no option does not look good. We could use the issue as a reference if we send and email to prod team or notify the ART team.
> 
The latest changes will add summary of blocking-secalerts in `Others` column. Something like : "Blocking Sec-laerts: 12345, 12346" and the advisory numbers are clickable hyperlinks to the corresponding advisory. I think this information can be used to send email to prod-sec team.

> On the other side creating a new command would have its own item in the check list (but this would break the current template).
> 
Agreed. One option could be to include this command in the existing table using the current template. However, that may not be very meaningful, as this command is not a primary action item in the z-stream release process. It simply serves as a check to ensure there are no blocking security alerts—similar to how we verify container grades. 

> I agree that checking the sec. alerts in the `change-advisory-status` is too late and we do it manually sooner and it would have deserved an earlier automatic check.
+1

With the latest changes Others column get updated with useful summary of blocking sec-alerts. If there are no blocking alerts then nothing is added to `Others` column. I think its better we keep the drop down with yes/no option to track the check is done for blocking sec-alerts or we can add this the to main table(Which probably may not be a right choice keeping release related tasks in mind). 
Another thing what we can do is stick to only `Others` column and updated the summary of security-alerts check like : 
1. No blocking alerts then `Others` column will have : "No blocking sec-alerts"
2. If there are blocking alerts then : "Blocking Sec-laerts: 12345, 12346"

WDYT?

rioliu-rh commented on oar/README.md:
````diff
@@ -106,3 +107,10 @@ $ oar -r $release-version drop-bugs
+```
+$ oar -r $release-version check-blocking-sec-alerts
+$ oar -r $release-version check-blocking-sec-alerts --check-only
+$ oar -r $release-version check-blocking-sec-alerts --output json
````
is there any other program will consume the output of this cmd, I think this option is useless

jhuttana replied:
Not in our tool but in case if we have to give detailed output to prod-sec team (If in special case) this could be useful for debugging I think

rioliu-rh commented on oar/core/const.py:
```diff
@@ -23,6 +23,8 @@
 LABEL_SIPPY = "A25"
 LABEL_SIPPY_MAIN = "A26"
 LABEL_SIPPY_AUTO_RELEASE = "A27"
+LABEL_BLOCKING_SEC_ALERTS = "A29"
+LABEL_BLOCKING_SEC_ALERTS_DROPDOWN = "B29"
```
the colum label of `others` should col H, also when we update this column need to get a blank cell to update, it's possible that there is CVP ticket is updated under column already

rioliu-rh commented on oar/core/worksheet.py:
```diff
@@ -616,6 +617,124 @@ def create_test_results_links(self):
+                    # Single advisory - format as: "🔴 Blocking Sec-Alert: [12345]"
+                    advisory = blocking_advisories[0]
+                    advisory_id = str(advisory["errata_id"])
+                    hyperlink = self._to_hyperlink(util.get_advisory_link(advisory_id), advisory_id)
```
this method cannot update multiple hyperlinks to single cell

jhuttana replied:
Hmm you are right. Probably I will have to use the customized method you have created.

rioliu-rh commented on oar/core/worksheet.py:
```diff
@@ -616,6 +617,124 @@ def create_test_results_links(self):
+                        advisory_id = str(advisory["errata_id"])
+                        hyperlink = self._to_hyperlink(util.get_advisory_link(advisory_id), advisory_id)
+                        hyperlinked_ids.append(hyperlink)
+                    summary_text = f"🔴 Blocking Sec-Alerts: {', '.join(hyperlinked_ids)}"
```
let's remove the special char like 🔴  from the code

rioliu-rh commented:
> ==================================================
> 📊 Summary:
>   • Total advisories: 4
>   • RHSA advisories checked: 2
>   • Blocking advisories found: 0
>   • Errors encountered: 0
> 
> ✅ No blocking security alerts found
> 
> 📋 Advisory Details:
>   ✅ 152071 (RHSA/image)
>   ✅ 152072 (RHSA/rpm)
>   ⏭️ 152073 (RHBA/extras) (RHBA advisories don't have security alerts)
>   ⏭️ 152074 (RHBA/metadata) (RHBA advisories don't have security alerts)
> 2025-07-23T15:08:21Z: INFO: Updating worksheet with blocking security alerts status...
> 📝 Worksheet Status:
>   • Value remains 'No' in cell B29 (no change needed)
> 
> ✅ Exit code 0: No blocking security alerts
> 2025-07-23T15:08:23Z: ERROR: check blocking security alerts failed
> Traceback (most recent call last):
>   File "/home/jhuttana/workarea/openShift_qe/release-tests/oar/cli/cmd_check_blocking_sec_alerts.py", line 193, in check_blocking_sec_alerts
>     ctx.exit(0)
>     ~~~~~~~~^^^
>   File "/home/jhuttana/.local/lib/python3.13/site-packages/click/core.py", line 699, in exit
>     raise Exit(code)
> click.exceptions.Exit: 0
> ❌ Error: 0
> ```

I think for output cmd, the func `has_blocking_secruity_alert` already covered this part, we don't need to create a new template, what we need from this cmd is, which RHSA has blocking secalert. let's make it simple.

rioliu-rh commented:
The `cmd_check_blocking_sec_alerts.py` can be significantly simplified since:
1. The core functionality is already provided by `Advisory.has_blocking_secruity_alert()`
2. The method already logs detailed alert information
3. The verbose report generation is redundant

Suggested simplified version would:
1. Just iterate through `RHSA` advisories
2. Call `has_blocking_secruity_alert()` on each
3. Collect blocking advisory IDs
4. Optionally update worksheet with basic status

This would reduce the file size by ~75% while maintaining all key functionality.

jhuttana commented:
Sample output after latest changes

1. Options available with the command
```
$ oar check-blocking-sec-alerts -h
Usage: oar check-blocking-sec-alerts [OPTIONS]

  Check for blocking security alerts across all advisories and update Others
  column with results

Options:
  --check-only  Only check for blocking alerts without updating worksheet
  -h, --help    Show this message and exit.

```
2. Testcase execution output
```
$ python -m unittest tests.test_worksheet.TestTestReport.test_blocking_sec_alerts -v
/home/jhuttana/.local/lib/python3.13/site-packages/gspread/auth.py:63: DeprecationWarning: [Deprecated][in version 6.0.0]: client_factory will be replaced by gspread.http_client types
  warnings.warn(
/home/jhuttana/.local/lib/python3.13/site-packages/gspread/client.py:588: DeprecationWarning: [Deprecated][in version 6.0.0]: this class will be deprecated and moved to gspread.http_client package
  warnings.warn(
test_blocking_sec_alerts (tests.test_worksheet.TestTestReport.test_blocking_sec_alerts)
Test blocking sec-alerts functionality in Others column ... ok

----------------------------------------------------------------------
Ran 1 test in 9.462s

OK
<img width="726" height="307" alt="image" src="https://github.com/user-attachments/assets/bea81e73-69a2-454a-b0ad-85ac03620971" />

```
3. Try it on 4.18.19 where there is already a CVP JIRA ticket present on the `Others` column to ensure the existing entry is not overwritten
```
$ oar -r 4.18.19 check-blocking-sec-alerts
2025-07-24T15:50:39Z: INFO: RHSA advisory 151421 does not have blocking secalert
2025-07-24T15:50:47Z: INFO: RHSA advisory 151422 does not have blocking secalert
No blocking security alerts found
2025-07-24T15:50:48Z: INFO: Added blocking sec-alerts status to Others column at row 9: OK: No Blocking Sec-Alerts
Worksheet updated

<img width="726" height="307" alt="image" src="https://github.com/user-attachments/assets/9d97fe7c-480a-4b2f-9002-fa7a2b6eb544" />

```

rioliu-rh commented on oar/core/notification.py:
```diff
@@ -313,7 +313,6 @@ def send_email(self, to_addrs, subject, content):
 
         logger.info(f"sent email to {to_addrs} with subject: <{subject}>")
 
-
```
need to revert changes in this file

rioliu-rh commented on oar/cli/cmd_check_blocking_sec_alerts.py:
```diff
@@ -0,0 +1,88 @@
+
+        # Simplified output - show only RHSA advisories with blocking security alerts
+        if has_blocking:
+            click.echo("BLOCKING SECURITY ALERTS FOUND:")
```
let's replace `click.echo` with `logger`

jhuttana commented:
> > 3. Other's column got updated properly in the test-report
> > 
> > <img alt="image" width="692" height="268" src="https://private-user-images.githubusercontent.com/60094631/471380317-3439e837-7788-49fe-af0c-ebcf05969165.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM2OTIwMzQsIm5iZiI6MTc1MzY5MTczNCwicGF0aCI6Ii82MDA5NDYzMS80NzEzODAzMTctMzQzOWU4MzctNzc4OC00OWZlLWFmMGMtZWJjZjA1OTY5MTY1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzI4VDA4MzUzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNlNjc4NmZlMGMyYTUwMTViMDQzNzhhMDRiMGE5ODFhNWQ2NTllMjIxZjhlNGI2YzZkZDNmZmRjYWI5MWI1NjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.o0vRXv_qNZr8g6fCSQGS7K8GaP_X99WtPFql87I0obA">
> 
> we need to avoid duplicate item for secalert, if the subsequent run found that secalert is fixed, we need to replace existing cell related to secalert with words like `OK: xxx`

We can do that but I think the subsequent updates will act as history for sec-alerts. WDYT?

rioliu-rh commented:
> > > 3. Other's column got updated properly in the test-report
> > > 
> > > <img alt="image" width="692" height="268" src="https://private-user-images.githubusercontent.com/60094631/471380317-3439e837-7788-49fe-af0c-ebcf05969165.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTM2OTIwMzQsIm5iZiI6MTc1MzY5MTczNCwicGF0aCI6Ii82MDA5NDYzMS80NzEzODAzMTctMzQzOWU4MzctNzc4OC00OWZlLWFmMGMtZWJjZjA1OTY5MTY1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzI4VDA4MzUzNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNlNjc4NmZlMGMyYTUwMTViMDQzNzhhMDRiMGE5ODFhNWQ2NTllMjIxZjhlNGI2YzZkZDNmZmRjYWI5MWI1NjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.o0vRXv_qNZr8g6fCSQGS7K8GaP_X99WtPFql87I0obA">
> > 
> > 
> > we need to avoid duplicate item for secalert, if the subsequent run found that secalert is fixed, we need to replace existing cell related to secalert with words like `OK: xxx`
> 
> We can do that but I think the subsequent updates will act as history for sec-alerts. WDYT?

I think we don't need the cmd history recorded in test report, it just shows the latest status of QE verification

rioliu-rh commented on oar/cli/cmd_check_blocking_sec_alerts.py:
```diff
@@ -0,0 +1,88 @@
+
+        # Check each advisory using existing has_blocking_secruity_alert method
+        for advisory in advisories:
+            advisory_info = {
```
`advisory_info` is useless, all the attribute can be retrieved from `Advisory` object.

rioliu-rh commented on oar/cli/cmd_check_blocking_sec_alerts.py:
```diff
@@ -0,0 +1,88 @@
+
+        # Update worksheet if not in check-only mode
+        if not check_only:
+            others_added = report.add_security_alert_status_to_others_section(has_blocking, details["blocking_advisories"])
+            if others_added:
```
```suggestion
            if report.add_security_alert_status_to_others_section(has_blocking, details["blocking_advisories"]):
```
